feat(perf-issues) Integrate detection into save transaction pipeline

udameli opened the pull request:
This PR updates the detection to return performance problems so they can be used for group creation/update. Only the output of duplicate span detector is used for now. This inserts performance problems into `job["performance_problems"]` to be used in a later pr.

Commit:
return perf problems from detection

k-fish commented on src/sentry/event_manager.py:
```diff
@@ -2009,15 +2018,16 @@ def save_transaction_events(jobs, projects):
     _derive_interface_tags_many(jobs)
     _calculate_span_grouping(jobs, projects)
+
+    # NEW CODEPATH HERE
+    # add here detection here
```
nit: comments needed?

Commit:
fix

udameli commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +147,64 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
+    detected_problems = [
+        (i, detector_type)
+        for detector_type, d in used_perf_issue_detectors.items()
+        for _, i in d.stored_issues
```
getting the second element here, since `stored_issues` is a dict with `fingerprint` and `perf problem`

k-fish replied:
We can just use a regular for loop here if you want, nested comprehensions are pretty confusing (my bad, lol)

k-fish commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -40,14 +65,16 @@ def detect_performance_issue(data: Event):
     except Exception as e:
         sentry_sdk.capture_exception(e)
+        # TODO is it better to return an empty list or None?
+        return None
```
Return type is List, so at least you'd want to change it to optional

k-fish commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +147,64 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
+
+def prepare_problem_for_grouping(problem: PerformanceSpanProblem, data: Event):
+    transaction_name = data.get("transaction")
+    first_span_id = problem[2][0]
```
Problem spans can be accessed by attr instead of index, and we should imo, having unnamed index lookups is usually surprising, it's not really a perf benefit to use list / index lookup here.

udameli replied:
👍 addressed

Commit:
return dataclass

k-fish commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +147,64 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
 
+    # Create performance issues for duplicate spans first
+    used_perf_issue_detectors = {
+        DetectorType.DUPLICATE_SPANS_HASH: DuplicateSpanHashDetector(detection_settings, data),
```
We can just pick one of these (we probably want `hash`), and move forward with just the one to reduce confusion, it's not like we can't re-add the other back later, mostly since we are now using `hash` in the _group fingerprint_. Unless you were thinking to check out the output of these vs each other? 🤔

udameli replied:
that makes sense 👍 picking Hash one

k-fish commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +147,64 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
+        (i, detector_type)
+        for detector_type, d in used_perf_issue_detectors.items()
+        for i in d.stored_issues
+    ]
```
nit: we can just use regular for loop, nested list comprehensions are a little hard to read (my bad, lol)

k-fish commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +147,64 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
+    detected_problems = [
+        (i, detector_type)
+        for detector_type, d in used_perf_issue_detectors.items()
+        for i in d.stored_issues
```
nit: we should probably change `stored_issues` on the detectors as well to be `problem` to make sure all the terminology is the same

k-fish commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +147,64 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
+    return prepared_problem
+
+
+def fingerprint_group(transaction_name, span_op, hash, problem_class):
+    signature = (str(transaction_name) + str(span_op) + str(hash) + str(problem_class)).encode(
+        "utf-8"
+    )
+    full_fingerprint = hashlib.sha1(signature).hexdigest()
+    return full_fingerprint
```
So if we have this fingerprint stored in an indexed way on a transaction row in clickhouse... we can test out the hit rate of `transaction_name` vs no-`transaction_name` very easily.

One other comment, I'd consider maybe taking the known cardinality strings and leave them as plain text, it will make looking at these values easier to interpret and be more human readable while we work on this (and queryable with string search if we want to do a bit of analysis). You can confirm the storage aspect w/ SnS, but it will remain a bounded-size string. 

eg. for `problem_class` since there are a known amount, we don't really need to collapse it into the the fingerprint.

```py
    signature = (str(transaction_name) + str(span_op) + str(hash)).encode(
        "utf-8"
    )
    full_fingerprint = hashlib.sha1(signature).hexdigest()
    return f"{problem_class}-{full_fingerprint}"
```

Instead of standing up a new column in theory you could prefix it with a version of our fingerprinting eg. 

```
f"1-{problem_class}-{full_fingerprint}"
```

which _does_ require a wildcard search on the groups, but for developing this feature (up until we potentially wipe that column before going live) it would let us test out different group fingerprint strategies pretty quickly by just incrementing a version to discern between changes (and using discover to check, say, transaction_name vs. no transaction_name)

udameli replied:
got it. opted in to using `f"1-{problem_class}-{full_fingerprint}"` 👍

k-fish replied:
Just checked, forgot we were doing `group_ids` on transaction now, not fingerprints, so this isn't searchable regardless, let's just leave it and figure out if we need to change it later, good enough for now.

k-fish reviewed (approved):
Nothing critical, just some comments, nice 👏

Commit:
address comments

Commit:
return an empty list on error from detection

Commit:
stored issues -> stored problems

Commit:
update fingerprint

Commit:
update tests

wedamija reviewed (commented):
This generally looks good to me, although I'm not sure if some changes are leaking through from the base branch?

It might actually be better to just base this off of master and have `rjo100/create-performance-groups/WOR-2100` use this as its base branch instead. Then we could get this merged sooner. We'd only call `_detect_performance_problems(jobs, projects)` in `save_transaction_events`. That won't have any effect without the `_save_aggregate_performance` call, so should be safe to do.

Up to you if you want to do this. Let me know what you decide and then I'll review this in a bit more detail

udameli commented:
ok I opened https://github.com/getsentry/sentry/pull/38308 instead

wedamija committed:
fix bug around `used_perf_issue_detectors`
add checks that performance problems are correct to the test

wedamija committed:
small tidy-up

wedamija committed:
add to choices

wedamija commented on src/sentry/types/issues.py:
```diff
@@ -17,6 +21,10 @@ class GroupCategory(Enum):
     GroupType.ERROR: GroupCategory.ERROR,
     GroupType.PERFORMANCE_N_PLUS_ONE: GroupCategory.PERFORMANCE,
     GroupType.PERFORMANCE_SLOW_SPAN: GroupCategory.PERFORMANCE,
+    GroupType.PERFORMANCE_SEQUENTIAL_SLOW_SPANS: GroupCategory.PERFORMANCE,
```
I think we'll need to figure out a way to centralize this better going forward, it's a bit painful atm to remember to update this in 3 places. We can probably improve it after EA/GA

wedamija committed:
Determine `group_type` dynamically in `prepare_problem_for_grouping`.
Add test for `prepare_problem_for_grouping`
Add test to make sure we have a mapping between all detector types and group types.

wedamija committed:
no need to filter none anymore

jjbayer commented on src/sentry/event_manager.py:
```diff
@@ -1897,6 +1898,12 @@ def _calculate_span_grouping(jobs, projects):
+@metrics.wraps("save_event.detect_performance_problems")
+def _detect_performance_problems(jobs, projects):
+    for job in jobs:
+        job["performance_problems"] = detect_performance_problems(job["data"])
```
Do we want to wrap this code in a `try...catch` block, similar to how `_calculate_span_grouping` does it?

wedamija replied:
`detect_performance_problems` already wraps its entire body in try/catch, I don't think we need to double up

udameli replied:
`detect_performance_problems` already has a `try catch` block. so if there was an exception, it'd just return an empty list. Do you think that's sufficient?

udameli commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +146,61 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
+) -> PerformanceProblem:
+    transaction_name = data.get("transaction")
+    spans_involved = problem.spans_involved
+    first_span_id = spans_involved[0]
```
yes, I think so 👍

udameli commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -120,12 +146,61 @@ def _detect_performance_issue(data: Event, sdk_span: Any):
+    first_span_id = spans_involved[0]
+    spans = data.get("spans", [])
+    first_span = next((span for span in spans if span["span_id"] == first_span_id), None)
+    op = first_span["op"]
+    hash = first_span["hash"]
+    desc = first_span["description"]
```
span.op and span.hash are guaranteed to exist. `visit_span` checks that in `settings_for_span(span)`

wedamija committed:
fix test to not need to be updated whenever we add a new type

Commit:
Merge branch 'master' into du/feat/integrate-detection-and-group-creation

Commit:
better typing for PerformanceProblemMap
